File: data/src/main/java/com/thinkzi/weather/data/executor/JobExecutor.java
package com.thinkzi.weather.data.executor;

import androidx.annotation.NonNull;

import com.thinkzi.weather.domain.executor.ThreadExecutor;

import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * provide an implementation of ThreadExecutor for thread pool to execute works
 */
@Singleton
public class JobExecutor implements ThreadExecutor {

    private final ThreadPoolExecutor _threadPoolExecutor;

    @Inject
    JobExecutor() {
        this._threadPoolExecutor = new ThreadPoolExecutor(3, 5, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new JobThreadFactory());
    }

    @Override
    public void execute(@NonNull Runnable command) {
        this._threadPoolExecutor.execute(command);
    }

    private static class JobThreadFactory implements ThreadFactory {
        private int counter = 0;


        @Override
        public Thread newThread(Runnable r) {
            return new Thread(r, "android_" + counter++);
        }
    }

}

File: data/src/main/java/com/thinkzi/weather/data/executor/JobExecutor.kt
package com.thinkzi.weather.data.executor

import com.thinkzi.weather.domain.executor.ThreadExecutor

import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadFactory
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

import javax.inject.Inject
import javax.inject.Singleton

/**
 * provide an implementation of ThreadExecutor for thread pool to execute works
 */
@Singleton
class JobExecutor @Inject
internal constructor() : ThreadExecutor {

    private val _threadPoolExecutor: ThreadPoolExecutor

    init {
        this._threadPoolExecutor = ThreadPoolExecutor(
            3,
            5,
            10,
            TimeUnit.SECONDS,
            LinkedBlockingQueue(),
            JobThreadFactory()
        )
    }

    override fun execute(command: Runnable) {
        this._threadPoolExecutor.execute(command)
    }

    private class JobThreadFactory : ThreadFactory {
        private var counter = 0


        override fun newThread(r: Runnable): Thread {
            return Thread(r, "android_" + counter++)
        }
    }

}
